File: 0931-minimum-falling-path-sum/0931-minimum-falling-path-sum.java
// class Solution {
//     public int solve(int [][]grid, int i, int j, int m, int[][] dp){
//         if(j<0 ||j>=m) return 1000000;
//         if(i==0) return grid[i][j];
//         if(dp[i][j]!=-1) return dp[i][j];
//         int top,topleft,topright;
        
//              top = grid[i][j] + solve(grid,i-1,j,m,dp);
//              topleft =  grid[i][j] + solve(grid,i-1,j-1,m,dp);
//              topright =  grid[i][j] + solve(grid,i-1,j+1,m,dp);
            
        
//         return dp[i][j] = Math.min(top,Math.min(topleft,topright));
//         // for(int k=-1;k<=1;k++){
//         //     int idx = j+k;
//         //     if(idx>=0 && idx<=m-1){
//         //         int temp = solve(grid, i-1, idx, m, dp);
//         //         ans = Math.min(ans, temp);
//         //     }
//         // }
//         // return dp[i][j] = ans + grid[i][j];
//     }
//     public int minFallingPathSum(int[][] matrix) {
//         int n = matrix.length;
//         int m = matrix[0].length;
//         if(n==1) return matrix[0][0];
//         int [][]dp = new int[n][m];
//         for(int i=0;i<n;i++){
//             Arrays.fill(dp[i],-1);
//         }
//         int ans = Integer.MAX_VALUE;
//         for(int i=0;i<m;i++){
//             ans = Math.min(ans, solve(matrix, n-1, i, m, dp));
//             // for(int k=-1;k<=1;k++){
//             //     int idx = i+k;
//             //     if(idx>=0 && idx<=m-1){
//             //        int temp = matrix[n-1][i] + solve(matrix,n-2,idx, m, dp);
//             //        ans = Math.min(ans, temp); 
//             //     }
//             // }
//         }
//         return ans;
//     }
// }

class Solution {
    public int minFallingPathSum(int[][] matrix) {
        int n = matrix.length;
        int[][] memo = new int[n][n]; 
        for (int i = 0; i < n; i++) {
            Arrays.fill(memo[i], Integer.MAX_VALUE);
        }

        int minSum = Integer.MAX_VALUE;
        for (int i = 0; i < n; i++) {
            minSum = Math.min(minSum, helper(matrix, 0, i, n, memo));
        }
        return minSum;
    }

    private int helper(int[][] matrix, int i, int j, int n, int[][] dp) {
        if (i == n - 1) {
            return matrix[i][j];
        }
        if (dp[i][j] != Integer.MAX_VALUE) {
            return dp[i][j];
        }
        
        int left = Integer.MAX_VALUE, right = Integer.MAX_VALUE;
        if (j > 0) left = matrix[i][j] + helper(matrix, i + 1, j - 1, n, dp);
        int down = matrix[i][j] + helper(matrix, i + 1, j, n, dp);
        if (j < n - 1) right = matrix[i][j] + helper(matrix, i + 1, j + 1, n, dp);

        return dp[i][j] = Math.min(Math.min(left, down), right);
    }
}

File: 0931-minimum-falling-path-sum/0931-minimum-falling-path-sum.cpp
class Solution {
public:
    int solve(vector<vector<int>>&mat, int i , int j,vector<vector<int>>&dp){
        int n = mat.size();
        int m = mat[0].size();

        
        if (j < 0 || j >= m) return 1e9;

        if (i == 0) return mat[0][j];
        
        if(dp[i][j]!= -1) return dp[i][j] ;
        
        int s = solve(mat, i - 1, j,dp);
        int lg = solve(mat, i - 1, j - 1,dp);
        int rd = solve(mat, i - 1, j + 1,dp);


        return dp[i][j] = mat[i][j] + min({s, lg, rd});

    }
    int minFallingPathSum(vector<vector<int>>& matrix) {
        int m = matrix.size();
        int n = matrix[0].size();
        vector<int> dp(n);
        int ans = 1e8;
        for(int i=0;i<n;i++){
          dp[i] = matrix[0][i];
        }
        for(int i=1;i<m;i++){
            vector<int> temp(n);
            for(int j=0;j<n;j++){
                int top=0;
                int topleft=1e8;
                int topright=1e8;
                top = matrix[i][j] + dp[j];
                if(j>0){
                    topleft = matrix[i][j] + dp[j-1];
                }
                if(j<n-1){
                    topright = matrix[i][j] + dp[j+1];
                }
                temp[j] = min(top,min(topleft,topright));
            }
            dp = temp;
        }
        for(int j=0;j<n;j++){
            ans = min(ans,dp[j]);
        }
        return ans;
//         int n = matrix.size() ;
//         int m = matrix[0].size() ;
//         vector<vector<int>>dp(n, vector<int>(m,-1)) ;

//         int mini = 1e9 ;
//         for(int j = 0 ;j<m ; j++){
//             mini = min(mini ,solve(matrix,n-1,j,dp)) ;
//         }

//         return mini ;
    }
};
